### ISSUE TITLE
[feature] only trace functions in specific files

### ISSUE THREAD

**Issue by Co1lin** [cool-RR/PySnooper#250]
It seems that the current tracer does not support only tracing files in some specific files. The user can of course filter out the results after execution, but I found that this affects efficiency. To trace a large project containing many functions simultaneously, it is hard to specify a fixed depth. Setting it too large will lead to diving into many builtin modules, which is very slow. Instead, using path/filenames as the filtering criteria is more controllable.

I'll implement this for my needs these days in my forked repo. If you guys think it's useful, I can submit a PR. But which style for the additional parameter is suitable? Welcome any comments, thanks!

```python
# usage: only lines in files under src_dir will be traced
@pysnooper.snoop(src_dir='myproj') # will be set as self.src_dirs = [src_dir]
@pysnooper.snoop(src_dirs=[ 'myproj1', 'myproj2', ])

# implementation draft
def _file_in_dir(file, directory):
    # https://stackoverflow.com/q/3812849/12388699
    directory = os.path.join(os.path.realpath(directory), '')
    file = os.path.realpath(file)
    return os.path.commonprefix([file, directory]) == directory

class Tracer:
    def trace(self, frame, event, arg):

        ### Checking whether we should trace this line: #######################
        #                                                                     #
        # We should trace this line either if it's in the decorated function,
        # or the user asked to go a few levels deeper and we're within that
        # number of levels deeper.
        if not any(_file_in_dir(frame.f_code.co_filename, d) for d in self.src_dirs):
            return None
        if not (frame.f_code in self.target_codes or frame in self.target_frames):
            if self.depth == 1:
                # We did the most common and quickest check above, because the
                # trace function runs so incredibly often, therefore it's
                # crucial to hyper-optimize it for the common case.
                return None
```

**Comment by cool-RR** [comment]
Good idea. Here's my proposed API. The argument will be called `source_paths` and will have a default value of `None`, indicating that there shouldn't be filtering. 

You could feed into it an iterable where every item can be a string or a `pathlib.Path` object. (Be mindful that you're not breaking Python 2.7 compatibility when checking for that.) These paths get canonicalized as you've done with `realpath` into a string. Then they get put in a tuple (because the user may have used any other kind of sequence or iterable.)

These paths may point to either a folder or a file. If it's a folder, then any file in that folder or its subfolders is included. If it's a file, then just that file.

Additionally, we can add boolean arguments `include_stdlib=True` and `include_site_packages=True`. Let's say that we don't allow these arguments to be used in conjuction with `source_paths`, and we raise an error if someone passes `False` into them while also passing something into `source_paths`. For these to work, we need to be sure we're correctly detecting the stdlib and site-packages folders.

What do you think?
